src/track/trackmetadatataglib.cpp

```diff
@@ -564,44 +590,104 @@ void writeTrackMetadataIntoTag(
     }
 
-    // Set the numeric year if available
-    const QDate yearDate(
-            TrackMetadata::parseDateTime(trackMetadata.getYear()).date());
-    if (yearDate.isValid()) {
-        pTag->setYear(yearDate.year());
+    // Specialized write functions for tags derived from Taglib::Tag might
+    // be able to write the complete string from trackMetadata.getYear()
+    // into the corresponding field. In this case parsing the year string
+    // here should be omitted.
+    if (0 == (writeMask & WRITE_TAG_OMIT_YEAR)) {
+        // Set the numeric year if available
+        const QDate yearDate(
+                TrackMetadata::parseDateTime(trackMetadata.getYear()).date());
+        if (yearDate.isValid()) {
+            pTag->setYear(yearDate.year());
+        }
+    }
+
+    // The numeric track number in TagLib::Tag does not reflect the total
+    // number of tracks! Specialized write functions for tags derived from
+    // Taglib::Tag might be able to handle both trackMetadata.getTrackNumber()
+    // and trackMetadata.getTrackTotal(). In this case parsing the track
+    // number string here is useless and should be omitted.
+    if (0 == (writeMask & WRITE_TAG_OMIT_TRACK_NUMBER)) {
+        // Set the numeric track number if available
+        TrackNumbers parsedTrackNumbers;
+        const TrackNumbers::ParseResult parseResult =
+                TrackNumbers::parseFromString(trackMetadata.getTrackNumber(), &parsedTrackNumbers);
+        if (TrackNumbers::ParseResult::VALID == parseResult) {
+            pTag->setTrack(parsedTrackNumbers.getActual());
+        }
+    }
+}
+
+void writeMP4Atom(
+        TagLib::MP4::Tag* pTag,
+        const TagLib::String& key,
+        const QString& str) {
+    if (str.isEmpty()) {
+        // Purge empty atoms
+        pTag->itemListMap().erase(key);
+    } else {
+        TagLib::StringList strList(toTagLibString(str));
+        pTag->itemListMap()[key] = std::move(strList);
     }
-    // Derived tags might be able to write the complete string
-    // from trackMetadata.getYear() into the corresponding field.
+}
 
-    bool trackNumberValid = false;
-    uint track = trackMetadata.getTrackNumber().toUInt(&trackNumberValid);
-    if (trackNumberValid && (track > 0)) {
-        pTag->setTrack(track);
+void writeAPEItem(
+        TagLib::APE::Tag* pTag,
+        const TagLib::String& key,
+        const TagLib::String& str) {
+    if (str.isEmpty()) {
+        // Purge empty items
+        pTag->removeItem(key);
+    } else {
+        const bool replace = true;
+        pTag->addValue(key, str, replace);
     }
 }
 
-template<typename T>
-inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
-        const T& value) {
-    pTag->itemListMap()[key] = value;
+bool readXiphCommentField(
+        const TagLib::Ogg::XiphComment& tag,
+        const TagLib::String& key,
+        QString* pValue = nullptr) {
+    if (tag.fieldListMap().contains(key)) {
```

Comment:
Ah, I see, you version fixes the implicit insert of an empty StringList. 
But now we have again a double map lookup and the .isEmpty() is gone. 

How about: 
```
FieldListMap::ConstIterator it = tag.fieldListMap().find(key);
if (it != tag.fieldListMap().end() && !it.second.isEmpty()) {
    if (nullptr != pValue) {
        *pValue = toQStringFirstNotEmpty(it.second);
...
```
Unfortunately "it.second" relies in the DO_NOT_DOCUMENT part of the taglib header. 
Any other ideas?